base/abstractarraymath.jl

```diff
@@ -487,11 +525,21 @@ julia> collect(eachslice(M, dims=2))
  [3, 6, 9]
 ```
 """
-@inline function eachslice(A::AbstractArray; dims)
-    length(dims) == 1 || throw(ArgumentError("only single dimensions are supported"))
-    dim = first(dims)
-    dim <= ndims(A) || throw(DimensionMismatch("A doesn't have $dim dimensions"))
-    inds_before = ntuple(d->(:), dim-1)
-    inds_after = ntuple(d->(:), ndims(A)-dim)
-    return (view(A, inds_before..., i, inds_after...) for i in axes(A, dim))
+@inline eachslice(A;dims) = _eachslice(A,dims)
+
+@inline _eachslice(A,dim::Integer) = _eachslice(A,(dim,))
+@inline function _eachslice(A::AbstractArray{N}, dims::NTuple{M,Integer}) where {N,M}
+    for (i,dim) in enumerate(dims)
+        1 <= dim <= N || throw(DimensionMismatch("A doesn't have dimension $dim"))
+        dim in dims[1:i-1] && throw(DimensionMismatch("$dims are not unique"))
+    end
+    iter = CartesianIndices(map(dim -> axes(A,dim), dims))
+    lookup = ntuple(dim -> findfirst(isequal(dim), dims), N)
+
+    # determine element type
+    I = Tuple{map((a,l) -> l === nothing ? typeof(a) : eltype(a), axes(A), lookup)...}
+    hasstride = IndexStyle(A) === IndexLinear() # has linear indexing if parent does
```

Comment:
Continuing the discussion from (the now outdated) https://github.com/JuliaLang/julia/pull/32310#discussion_r540350659, the resulting subarray will only be linear if the slices are all across contiguous dimensions.  Here's a counter-example test-case:

```Julia
julia> A = rand(5,4,3,2); dims=(2,4); N=4; M=2;

julia> typeof(view(A, :, 1, :, 1))
SubArray{Float64,2,Array{Float64,4},Tuple{Base.Slice{Base.OneTo{Int64}},Int64,Base.Slice{Base.OneTo{Int64}},Int64},false}
```

This view is strided but not linear.

Reply:
That's unfortunate.

How bad would it be to use `Core.Compiler.return_type` here?

Reply:
It would certainly simplify things - assuming it doesn't make the compiler crash or compilation invalid?